File: holoken-app/src/main/java/com/holokenmod/ui/main/KeyPadFragment.java
package com.holokenmod.ui.main;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TableLayout;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import com.google.android.material.button.MaterialButton;
import com.holokenmod.R;
import com.holokenmod.databinding.KeyPadFragmentBinding;
import com.holokenmod.game.Game;
import com.holokenmod.options.ApplicationPreferences;
import com.holokenmod.options.CurrentGameOptionsVariant;
import com.holokenmod.options.DigitSetting;
import com.holokenmod.ui.GridCreationListener;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class KeyPadFragment extends Fragment implements GridCreationListener {
    private final List<MaterialButton> numbers = new ArrayList<>();
    private TableLayout controlKeypad;
    private Game game;
    
    private KeyPadFragmentBinding binding;
    
    public KeyPadFragment() {
        super(R.layout.key_pad_fragment);
    }
    
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup parent, Bundle savedInstanceState) {
        binding = KeyPadFragmentBinding.inflate(inflater, parent, false);
        
        return binding.getRoot();
    }
    
    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
    
    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        numbers.add(binding.button1);
        numbers.add(binding.button2);
        numbers.add(binding.button3);
        numbers.add(binding.button4);
        numbers.add(binding.button5);
        numbers.add(binding.button6);
        numbers.add(binding.button7);
        numbers.add(binding.button8);
        numbers.add(binding.button9);
        numbers.add(binding.button10);
        numbers.add(binding.button11);
        numbers.add(binding.button12);
        
        this.controlKeypad = binding.controls;
        
        for (final MaterialButton numberButton : numbers) {
            addButtonListeners(numberButton);
        }
        
        if (game != null) {
            setButtonLabels();
            setButtonVisibility();
        }
    }
    
    private void addButtonListeners(MaterialButton numberButton) {
        numberButton.setOnClickListener(v -> {
            final int d = Integer.parseInt(((MaterialButton) v).getText().toString());
            game.enterPossibleNumber(d);
        });
        numberButton.setOnLongClickListener(v -> {
            final int d = Integer.parseInt(((MaterialButton) v).getText().toString());
            game.enterNumber(d, ApplicationPreferences.getInstance().removePencils());
            
            return true;
        });
    }
    
    
    public void freshGridWasCreated() {
        if (!isAdded()) {
            return;
        }
        
        requireActivity().runOnUiThread( () -> {
            setButtonLabels();
            setButtonVisibility();
        });
    }
    
    private void setButtonLabels() {
        DigitSetting digitSetting = CurrentGameOptionsVariant.getInstance().getDigitSetting();
        
        Iterator<Integer> digits = digitSetting.getAllNumbers().iterator();
        
        if (digitSetting.containsZero()) {
            digits.next();
        }
        
        int visibleRows = ((int) Math.ceil(game.getGrid().getPossibleDigits().size() / 3.0));
        int lastVisibleNumber = visibleRows * 3 - 1;
        int i = 0;
        
        for (final Button numberButton : numbers) {
            int digit;
            
            if (i == lastVisibleNumber && digitSetting.containsZero()) {
                digit = 0;
            } else {
            digit = digits.next();
            }
            
            numberButton.setText(Integer.toString(digit));
            numberButton.setVisibility(i <= lastVisibleNumber ? View.VISIBLE : View.GONE);
            
            i++;
        }
    }
    
    private void setButtonVisibility() {
        for (MaterialButton number : numbers) {
            number.setEnabled(game.getGrid().getPossibleDigits().contains(Integer.parseInt(number.getText().toString())));
        }
        
        this.controlKeypad.setVisibility(View.VISIBLE);
    }
    
    public void setGame(Game game) {
        this.game = game;
        
        freshGridWasCreated();
    }
}

File: holoken-app/src/main/kotlin/com/holokenmod/ui/main/KeyPadFragment.kt
package com.holokenmod.ui.main

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.google.android.material.button.MaterialButton
import com.holokenmod.R
import com.holokenmod.databinding.KeyPadFragmentBinding
import com.holokenmod.game.Game
import com.holokenmod.options.ApplicationPreferences
import com.holokenmod.options.CurrentGameOptionsVariant
import com.holokenmod.ui.GridCreationListener
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import kotlin.math.ceil

class KeyPadFragment : Fragment(R.layout.key_pad_fragment), GridCreationListener, KoinComponent {
    private val game: Game by inject()
    private val applicationPreferences: ApplicationPreferences by inject()

    private lateinit var binding: KeyPadFragmentBinding
    private val numbers = mutableListOf<MaterialButton>()

    override fun onCreateView(
        inflater: LayoutInflater,
        parent: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = KeyPadFragmentBinding.inflate(inflater, parent, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        numbers += binding.button1
        numbers += binding.button2
        numbers += binding.button3
        numbers += binding.button4
        numbers += binding.button5
        numbers += binding.button6
        numbers += binding.button7
        numbers += binding.button8
        numbers += binding.button9
        numbers += binding.button10
        numbers += binding.button11
        numbers += binding.button12
                
        numbers.forEach {
            addButtonListeners(it)
        }

        setButtonLabels()
        setButtonVisibility()
    }

    private fun addButtonListeners(numberButton: MaterialButton) {
        numberButton.setOnClickListener {
            val d = numberButton.text.toString().toInt()
            game.enterPossibleNumber(d)
        }
        numberButton.setOnLongClickListener {
            val d = numberButton.text.toString().toInt()
            game.enterNumber(d, applicationPreferences.removePencils())
            true
        }
    }

    override fun freshGridWasCreated() {
        if (!isAdded) {
            return
        }
        requireActivity().runOnUiThread {
            setButtonLabels()
            setButtonVisibility()
        }
    }

    private fun setButtonLabels() {
        val digitSetting = CurrentGameOptionsVariant.instance.digitSetting
        val digits = digitSetting.allNumbers.iterator()

        if (digitSetting.containsZero()) {
            digits.next()
        }

        val visibleRows = ceil(game.grid.possibleDigits.size / 3.0).toInt()
        val lastVisibleNumber = visibleRows * 3 - 1

        var i = 0

        numbers.forEach {
            val digit: Int = if (i == lastVisibleNumber && digitSetting.containsZero()) {
                0
            } else {
                digits.next()
            }
            it.text = digit.toString()
            it.visibility = if (i <= lastVisibleNumber) View.VISIBLE else View.GONE
            i++
        }
    }

    private fun setButtonVisibility() {
        numbers.forEach {
            it.isEnabled = game.grid.possibleDigits.contains(it.text.toString().toInt())
        }
        binding.controls.visibility = View.VISIBLE
    }

    fun setGame(game: Game) {
        freshGridWasCreated()
    }
}